csharp/ql/src/semmle/code/csharp/dataflow/SignAnalysis.qll

```diff
@@ -0,0 +1,595 @@
+/**
+ * Provides sign analysis to determine whether expression are always positive
+ * or negative.
+ *
+ * The analysis is implemented as an abstract interpretation over the
+ * three-valued domain `{negative, zero, positive}`.
+ */
+
+private import csharp
+private import Ssa
+private import SsaReadPosition
+private import ConstantUtils
+private import RangeUtils
+private import semmle.code.csharp.controlflow.Guards as G
+private import Linq.Helpers as Linq
+
+private class Guard = G::Guard;
+
+private newtype TSign =
+  TNeg() or
+  TZero() or
+  TPos()
+
+private class Sign extends TSign {
+  string toString() {
+    result = "-" and this = TNeg()
+    or
+    result = "0" and this = TZero()
+    or
+    result = "+" and this = TPos()
+  }
+
+  Sign inc() {
+    this = TNeg() and result = TNeg()
+    or
+    this = TNeg() and result = TZero()
+    or
+    this = TZero() and result = TPos()
+    or
+    this = TPos() and result = TPos()
+  }
+
+  Sign dec() { result.inc() = this }
+
+  Sign neg() {
+    this = TNeg() and result = TPos()
+    or
+    this = TZero() and result = TZero()
+    or
+    this = TPos() and result = TNeg()
+  }
+
+  Sign bitnot() {
+    this = TNeg() and result = TPos()
+    or
+    this = TNeg() and result = TZero()
+    or
+    this = TZero() and result = TNeg()
+    or
+    this = TPos() and result = TNeg()
+  }
+
+  Sign add(Sign s) {
+    this = TZero() and result = s
+    or
+    s = TZero() and result = this
+    or
+    this = s and this = result
+    or
+    this = TPos() and s = TNeg()
+    or
+    this = TNeg() and s = TPos()
+  }
+
+  Sign mul(Sign s) {
+    result = TZero() and this = TZero()
+    or
+    result = TZero() and s = TZero()
+    or
+    result = TNeg() and this = TPos() and s = TNeg()
+    or
+    result = TNeg() and this = TNeg() and s = TPos()
+    or
+    result = TPos() and this = TPos() and s = TPos()
+    or
+    result = TPos() and this = TNeg() and s = TNeg()
+  }
+
+  Sign div(Sign s) {
+    result = TZero() and this = TZero() and s = TNeg()
```

Comment:
I change this from `result = TZero() and s = TNeg()`. I think we need the extra constraint on `this`. If this is really the case, then we should fix both the Java and C++ implementation.

I made the same change in the next line too.

Reply:
For ints, `3 / 5 = 0`. So `this` doesn't have to be zero in order to get a zero result.

Reply:
Ahh, thanks. I couldn't figure it out why that was missing.